File: webapp/src/license.ts
import {GlobalState} from 'mattermost-redux/types/store';
import {getLicense, getConfig} from 'mattermost-redux/selectors/entities/general';

// isE20LicensedOrDevelopment returns true when the server is licensed with a Mattermost
// Enterprise E20 License, or has `EnableDeveloper` and `EnableTesting` configuration settings
// enabled, signaling a non-production, developer mode.
export const isE20LicensedOrDevelopment = (state: GlobalState): boolean => {
    const license = getLicense(state);

    // Use the presence of a known e20 feature as a check to determine licensing.
    if (license?.MessageExport === 'true') {
        return true;
    }

    return isConfiguredForDevelopment(state);
};

// isE10LicensedOrDevelopment returns true when the server is licensed with a Mattermost
// Enterprise E10 License, or has `EnableDeveloper` and `EnableTesting` configuration settings
// enabled, signaling a non-production, developer mode.
export const isE10LicensedOrDevelopment = (state: GlobalState): boolean => {
    const license = getLicense(state);

    // Use the presence of a known E10 feature as a check to determine licensing.
    if (license?.LDAP === 'true') {
        return true;
    }

    return isConfiguredForDevelopment(state);
};

export const isConfiguredForDevelopment = (state: GlobalState): boolean => {
    const config = getConfig(state);

    return config.EnableTesting === 'true' && config.EnableDeveloper === 'true';
};

export const isEnterpriseEdition = (state: GlobalState): boolean => {
    const config = getConfig(state);

    return config.BuildEnterpriseReady === 'true';
};

export const isCloud = (state: GlobalState): boolean => {
    const license = getLicense(state);

    return license?.Cloud === 'true';
};

export const isPricingPlanDifferentiationEnabled = (state: GlobalState): boolean => {
    if (isCloud(state)) {
        return false;
    }

    if (!isEnterpriseEdition(state)) {
        return false;
    }

    return true;
};

File: server/enterprise/license.go
package enterprise

import (
	pluginapi "github.com/mattermost/mattermost-plugin-api"
)

type LicenseChecker struct {
	pluginAPIClient *pluginapi.Client
}

func NewLicenseChecker(pluginAPIClient *pluginapi.Client) *LicenseChecker {
	return &LicenseChecker{
		pluginAPIClient,
	}
}

// isAtLeastE10Licensed returns true when the server either has at least an E10 license or is configured for development.
func (e *LicenseChecker) isAtLeastE10Licensed() bool {
	config := e.pluginAPIClient.Configuration.GetConfig()
	license := e.pluginAPIClient.System.GetLicense()

	return pluginapi.IsE10LicensedOrDevelopment(config, license)
}

// isAtLeastE20Licensed returns true when the server either has at least an E20 license or is configured for development.
func (e *LicenseChecker) isAtLeastE20Licensed() bool {
	config := e.pluginAPIClient.Configuration.GetConfig()
	license := e.pluginAPIClient.System.GetLicense()

	return pluginapi.IsE20LicensedOrDevelopment(config, license)
}

// RetrospectiveAllowed returns true if the retrospective feature is allowed with the current license.
func (e *LicenseChecker) RetrospectiveAllowed() bool {
	return e.isAtLeastE10Licensed()
}

// TimelineAllowed returns true if the timeline feature is allowed with the current license.
func (e *LicenseChecker) TimelineAllowed() bool {
	return e.isAtLeastE10Licensed()
}

// StatsAllowed returns true if the stats feature is allowed with the current license.
func (e *LicenseChecker) StatsAllowed() bool {
	return e.isAtLeastE20Licensed()
}
